extensions/sql-database-projects/src/test/project.test.ts

```diff
@@ -707,26 +707,117 @@ describe('Project: sqlproj content operations', function (): void {
+	});
+
+	it('Intermediate folders for file should be automatically added to project', async function (): Promise<void> {
+		// Create new sqlproj
+		projFilePath = await testUtils.createTestSqlProjFile(baselines.newProjectFileBaseline);
+		const projectFolder = path.dirname(projFilePath);
+
+		// Create file under nested folders structure
+		const newFile = path.join(projectFolder, 'foo', 'bar', 'test.sql');
+		await fs.mkdir(path.dirname(newFile), { recursive: true });
+		await fs.writeFile(newFile, '');
+
+		// Open empty project
+		let project: Project = await Project.openProject(projFilePath);
+
+		// Add a file to the project
+		await project.addToProject([Uri.file(newFile)]);
+
+		// Validate that intermediate folders were added to the project
+		should(project.files.length).equal(3, 'Three entries are expected in the project');
+		should(project.files.map(f => ({ type: f.type, relativePath: f.relativePath })))
+			.containDeep([
+				{ type: EntryType.Folder, relativePath: 'foo\\' },
+				{ type: EntryType.Folder, relativePath: 'foo\\bar\\' },
+				{ type: EntryType.File, relativePath: 'foo\\bar\\test.sql' }]);
+	});
+
+	it('Intermediate folders for folder should be automatically added to project', async function (): Promise<void> {
+		// Create new sqlproj
+		projFilePath = await testUtils.createTestSqlProjFile(baselines.newProjectFileBaseline);
+		const projectFolder = path.dirname(projFilePath);
+
+		// Create nested folders structure
+		const newFolder = path.join(projectFolder, 'foo', 'bar');
+		await fs.mkdir(newFolder, { recursive: true });
+
+		// Open empty project
+		let project: Project = await Project.openProject(projFilePath);
+
+		// Add a file to the project
+		await project.addToProject([Uri.file(newFolder)]);
+
+		// Validate that intermediate folders were added to the project
+		should(project.files.length).equal(2, 'Two entries are expected in the project');
+		should(project.files.map(f => ({ type: f.type, relativePath: f.relativePath })))
+			.containDeep([
+				{ type: EntryType.Folder, relativePath: 'foo\\' },
+				{ type: EntryType.Folder, relativePath: 'foo\\bar' }]);
+	});
+
+	it('Should not add duplicate intermediate folders to project', async function (): Promise<void> {
+		// Create new sqlproj
+		projFilePath = await testUtils.createTestSqlProjFile(baselines.newProjectFileBaseline);
+		const projectFolder = path.dirname(projFilePath);
+
+		// Create file under nested folders structure
+		const newFile = path.join(projectFolder, 'foo', 'bar', 'test.sql');
+		await fs.mkdir(path.dirname(newFile), { recursive: true });
+		await fs.writeFile(newFile, '');
+
+		const anotherNewFile = path.join(projectFolder, 'foo', 'bar', 'test2.sql');
+		await fs.writeFile(anotherNewFile, '');
+
+		// Open empty project
+		let project: Project = await Project.openProject(projFilePath);
+
+		// Add a file to the project
+		await project.addToProject([Uri.file(newFile)]);
+		await project.addToProject([Uri.file(anotherNewFile)]);
```

Comment:
also add a case for when an intermediate folder gets added for a file, but the intermediate folder is also specified later in the array passed to `project.addToProject()`

Reply:
Good idea, let me add that!

Reply:
So there is an interesting behavior in that case - unless you use `doNotThrowOnDuplicate` flag, you will get `A folder with the path 'foo/bar' has already been added to the project` error, which looks reasonable, as that is the same behavior as if you were to add them one after another.

To be honest, I think we should not throw on duplicate records at all, ever. The goal of the caller to have the item in the project, if item is already there - goal achieved and we bail out with no modifications. What was the idea behind throwing an error when caller requests to add an item to the project and item is already present? What would be the scenario that cares about the error in this case?

Reply:
eventually, the ability to add existing files to a project through the UI will be added. If a user tries to add a file or folder that is already a part of the project, then there should be a notification that it wasn't added again because it's already a part of the project(like in VS). I guess in that case, it could just show the notification and not throw an error.

Reply:
What do you mean saying "like in VS"? Full-blown Visual Studio with SSDT does not show any messages - it just silently continues, if you try to add the same file. I think that is reasonable behavior, as ultimate goal for the user is to have file in the project. If it's already there - it just becomes no-op. Why any messages at all?

Reply:
oh never mind, I must've been thinking of the error that gets shown when adding duplicate database references. In that case I don't mind not throwing an error anymore.

Reply:
Alright, let me get rid of the flag and update the code. I'll see how complex the `addFolderItem` ends up being and decide whether I want a separate helper or just embed the logic in there and call it from `addScriptItem`.

Reply:
@kisantia, @Benjin  While working on the change, I noticed that `addScriptItem` currently allows content to be passed in, so that file can be created within the method. The current logic is strange and I'm not sure, if that was the intention or not:
1) If there is an entry in the sqlproj for the file, then error is thrown, content is not stored anywhere. This is the error we plan to remove.
2) If there is no entry in the sqlproj for the file, then contents will be written to a file and file will be added. Based on nodejs docs, this seems to also overwrite existing file without any prompts. Is this a bug or is this expected? I don't see any tests for this scenario, so not sure, if this is expected.

The reason why it looks strange to me is because behavior is different, based on the presence of the entry in sqlproj.
Now, if we remove the check for the presence of a file entry in sqlproj, then we have the following:
1) If there is an entry, then it will be returned, contents of the file are not modified. This sounds weird to me, as caller's intention was to store given content in the file with the given name. If the (2) was indeed a desired behavior and not a bug, then I think we should silently overwrite the file, like we are doing currently, if there is no entry.
2) If there is no entry in the sqlproj, then we again just write the contents to the file, no matter it exists or not and return new entry.

So should we just overwrite the file every time, if content was provided and return existing or new entry?

Reply:
Hm... actually it seems like we want to throw `constants.fileAlreadyExists`, if file already exists. Let me update the code to actually do that, so we will end up with:

0) If content is provided and file already exists - throw error, do not modify sqlproj or the file.
1) If there is an entry in sqlproj, then we create a file (if `content` is provided) and return existing entry.
2) If there is no entry in sqlproj, then we create a file (if `content` is provided) and add new entry.

Does this sound good?